### ISSUE TITLE
TextField maxLines change causes text containing newline chars overflow 

### ISSUE THREAD

**Issue by huntrist** [flutter/flutter#88422]
description:

build a TextField with maxLines = 4 when focused, and maxLines = 1 when unfocused;
when copy text with newline chars into a TextField (test input 1), then hit enter, text will overflow outside of the input area (expecting hide testline2, and testline3)

focused: (maxLines:4)
![97c64018c6cdcdc7e263be49464fc75](https://user-images.githubusercontent.com/27284528/129873476-271e00af-4578-45f3-8600-c142f65edc33.png)
unfocused: (maxLines:1)
![d421410428a163b9f60d8d1b15968f6](https://user-images.githubusercontent.com/27284528/129873499-38d37311-b7cd-4306-a8d4-ba7c9783666d.png)

but when put text with NO newline chars (test input 2), then hit enter, no overflow occures.
focused: (maxLines:4)
![df9b5227c988a8d5c5d606a95673586](https://user-images.githubusercontent.com/27284528/129873536-d6dbc9c0-a801-46c6-8126-21b53a0c55d4.png)
unfocused: (maxLines:1)
![b68f5bb3ea42e17535df831f7bf3759](https://user-images.githubusercontent.com/27284528/129873557-9e55ca57-f209-49cd-9047-681f70cae46b.png)

test input 1:
```
testline1
testline2
testline3
```

test input 2:

```
verylongtestline11111111111111111111111111111.test22222222222222222222test33333333333333
```

demo code:
```
void main() {
  runApp(App());
} 
 
class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Body(),
      ),
    );
  }
}

class Body extends StatefulWidget {

  @override
  BodyState createState() => BodyState();
}

class BodyState extends State<Body> {
  final FocusNode focusNode = FocusNode();
  final TextEditingController controller = TextEditingController();
  int maxLines = 1;

  @override
  void initState() {
    super.initState();
    focusNode.addListener(() {
      if (mounted) {
        print("focus changed!");
        if (focusNode.hasFocus) {
          maxLines = 4;
        } else {
          maxLines = 1;
        }
        setState((){});
      }
    });
  }

  @override
  Widget build(BuildContext context) {

    return SafeArea(
      child: Center(
        child: Container(
          color: ColorSet.red,
          child: TextField(
            onSubmitted: (String val) {
              focusNode.unfocus();
            },
            focusNode: focusNode,
            controller: controller,
            style: TextStyle(
              color: Colors.black,
              fontSize: 14,
            ),
            maxLines: maxLines,
            keyboardType: TextInputType.text,
            decoration: InputDecoration(
              hintText: "enter text",
              hintStyle: TextStyle(
                color: Colors.black,
                fontSize: 14,
                textBaseline: TextBaseline.alphabetic,
              ),
            ),
          ),
        ),
      ),
    );
  }
}
```

flutter doctor:
```
flutter doctor -v
[√] Flutter (Channel stable, 2.2.3, on Microsoft Windows [Version 10.0.19042.1165], locale zh-CN)
    • Flutter version 2.2.3 at D:\Env\flutter
    • Framework revision f4abaa0735 (7 weeks ago), 2021-07-01 12:46:11 -0700
    • Engine revision 241c87ad80
    • Dart version 2.13.4
    • Pub download mirror https://pub.flutter-io.cn
    • Flutter download mirror https://storage.flutter-io.cn

[!] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at D:\Env\Android\SDK
    • Platform android-30, build-tools 29.0.2
    • ANDROID_HOME = D:\Env\Android\SDK
    • Java binary at: D:\Env\Java\OpenJDK\bin\java.exe
    • Java version openjdk version "11.0.4" 2019-07-16
    X Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/windows#android-setup for more details.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/windows#android-setup for detailed instructions).

[√] Connected device (3 available)
    • sdk gphone x86 (mobile) • emulator-5554 • android-x86    • Android 11 (API 30) (emulator)
    • Chrome (web)            • chrome        • web-javascript • Google Chrome 92.0.4515.131
    • Edge (web)              • edge          • web-javascript • Microsoft Edge 91.0.864.41
```

**Comment by darshankawar** [comment]
@huntrist 
Check if [this](https://github.com/flutter/flutter/issues/71456) open issue sounds similar to yours.

**Comment by huntrist** [comment]
> @huntrist
> Check if [this](https://github.com/flutter/flutter/issues/71456) open issue sounds similar to yours.

might be related, but not sure it is the same problem I'm describing. 
What I'm describing is NOT text overflow ellipsis issue, but mainly new line char (\n) causes text being rendered OUTSIDE of the widget issue (as seen in the second image).
The situation in the second image looks like a bug to me. It should either show ellipsis (or has other TextOverflow options), or being put into one line and directly clip off - as in 4th image, or directly hide 2nd and 3rd line; but certainly NOT render outside of the widget.

**Comment by darshankawar** [comment]
Thanks for clarification. I tried the same on latest master (2.5.0-7.0.pre.154) but wasn't able to replicate it. Can you try the same and confirm the behavior ?

**Comment by huntrist** [comment]
> Thanks for clarification. I tried the same on latest master (2.5.0-7.0.pre.154) but wasn't able to replicate it. Can you try the same and confirm the behavior ?

sorry for the very late response, 
I tried it on the Stable 2.5.0 version, the problem still exists (exactly as the original post)
also tried it on Master 2.6.0-1.0.pre.219, the problem still exists (exactly as the original post)

**Comment by darshankawar** [comment]
Thanks for coming back. I was finally able to replicate it on latest master and stable on Android.

<details>
<summary> flutter doctor -v </summary>

```
[√] Flutter (Channel master, 2.6.0-1.0.pre.237, on Microsoft Windows [Version 10.0.19042.1165], locale en-US)
    • Flutter version 2.6.0-1.0.pre.237 at D:\FlutterSDK\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision a54b10254a (2 hours ago), 2021-09-10 07:22:02 +0300
    • Engine revision 57bc7414b8
    • Dart version 2.15.0 (build 2.15.0-92.0.dev)

[√] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at C:\Users\Win\AppData\Local\Android\sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[!] Visual Studio - develop for Windows (Visual Studio Community 2019 16.11.2)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.11.31624.102
    X The current Visual Studio installation is incomplete. Please reinstall Visual Studio.

[√] Android Studio (version 2020.3)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[√] VS Code (version 1.60.0)
    • VS Code at C:\Users\Win\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension can be installed from:
       https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[√] Connected device (4 available)
    • SM A260G (mobile) • 5200763ebcfa861f • android-arm    • Android 8.1.0 (API 27)
    • Windows (desktop) • windows          • windows-x64    • Microsoft Windows [Version 10.0.19042.1165]
    • Chrome (web)      • chrome           • web-javascript • Google Chrome 93.0.4577.63
    • Edge (web)        • edge             • web-javascript • Microsoft Edge 93.0.961.38

! Doctor found issues in 1 category.

[√] Flutter (Channel stable, 2.5.0, on Microsoft Windows [Version 10.0.19042.1165], locale en-US)
    • Flutter version 2.2.3 at D:\FlutterSDK\flutter
    • Framework revision f4abaa0735 (31 hours ago), 2021-09-07 12:46:11 -0700
    • Engine revision f0826da7ef
    • Dart version 2.14.0

[√] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at C:\Users\Win\AppData\Local\Android\sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[X] Visual Studio - develop for Windows
    X Visual Studio not installed; this is necessary for Windows development.
      Download at https://visualstudio.microsoft.com/downloads/.
      Please install the "Desktop development with C++" workload, including all of its default components

[√] Android Studio
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • android-studio-dir = C:\Program Files\Android\Android Studio
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[√] Connected device (5 available)
    • SM A260G (mobile) • 5200763ebcfa861f • android-arm    • Android 8.1.0 (API 27)
    • SM G975F (mobile) • RZ8M802WY0X      • android-arm64  • Android 11 (API 30)
    • Windows (desktop) • windows          • windows-x64    • Microsoft Windows [Version 10.0.19042.1165]
    • Chrome (web)      • chrome           • web-javascript • Google Chrome 93.0.4577.63
    • Edge (web)        • edge             • web-javascript • Microsoft Edge 92.0.902.84

! Doctor found issues in 1 category.

C:\Users\Win>

```

</details>

**Comment by minhnguyenandpad** [comment]
Hi, did you solve the problem. If you had please help me know how to handle it. Appreciated

**Comment by camsim99** [comment]
Has anyone been able to reproduce this on iOS? Wondering if this is Android-specific or not.

**Comment by LongCatIsLooong** [comment]
I think this is fixed by #112029
